routing/triert/table.go

```diff
@@ -0,0 +1,228 @@
+package triert
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+
+	// "github.com/libp2p/go-libp2p-xor/kademlia"
+	tkey "github.com/libp2p/go-libp2p-xor/key"
+	"github.com/libp2p/go-libp2p-xor/trie"
+	"github.com/plprobelab/go-kademlia/key"
+	"github.com/plprobelab/go-kademlia/network/address"
+)
+
+// TODO: move to key package and replace ErrInvalidKey
+// ErrKeyWrongLength indicates a key has the wrong length
+var ErrKeyWrongLength = errors.New("key has wrong length")
+
+// TrieRT is a routing table backed by a Xor Trie which offers good scalablity and performance
+// for large networks. All exported methods are safe for concurrent use.
+type TrieRT struct {
+	self key.KadKey
+
+	// keymu is held during keys and keyNodes mutations to serialize changes
+	keymu sync.Mutex
+
+	// keys holds a pointer to an immutable trie
+	// any store to keys must be performed while holding keymu, loads may be performed without the lock
+	keys atomic.Value
+
+	// keyNodes holds a mapping of kademlia key to node id (map[string]address.NodeID)
+	// note: this could be eliminated if the trie was modified to allow a tuple to be stored at each node
+	// any store to keyNodes must be performed while holding keymu, loads may be performed without the lock
+	keyNodes atomic.Value
+}
+
+func New(self key.KadKey) *TrieRT {
+	rt := &TrieRT{
+		self: self,
+	}
+	rt.keys.Store(&trie.Trie{})
+	rt.keyNodes.Store(map[string]address.NodeID{})
+	return rt
+}
+
+func (rt *TrieRT) Self() key.KadKey {
+	return rt.self
+}
+
+// AddPeer tries to add a peer to the routing table
+func (rt *TrieRT) AddPeer(ctx context.Context, node address.NodeID) (bool, error) {
+	kk := node.Key()
+	if kk.Size() != rt.self.Size() {
+		return false, ErrKeyWrongLength
```

Comment:
`ErrKeyWrongLength` could be defined in the package `routing` as it is a generic error for operations on routing tables.

Reply:
Sorry, I missed
```go
// TODO: move to key package and replace ErrInvalidKey
// ErrKeyWrongLength indicates a key has the wrong length
var ErrKeyWrongLength = errors.New("key has wrong length")
```
moving the error to the `key` package also makes sense.

Reply:
Yes, I didn't want to do this without discussion. I think we should replace `ErrInvalidKey` by `ErrKeyWrongLength` for two reasons:

1. a key of a different length isn't an invalid key, it's just the wrong length so the error is clearer
2. ErrInvalidKey isn't a single error so it's currently hard to use with `errors.Is`. We could improve that with a custom error type but I think it's simpler to have a specific error to signal incompatible lengths
3.

Reply:
I agree that it would make more sense to replace `ErrInvalidKey` with something like `ErrWrongKeyLength`.

I don't have much experience with defining error types in Go, all suggestions are welcome 😄  In error testing, I mostly used `require.Equal(t, ExpectedError, err)`, but I am not sure that it is the best practice

Reply:
Currently I'm using `ErrMismatchedKeyLength` from the `key/trie` package